mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java

```diff
@@ -0,0 +1,110 @@
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+public class CassandraMessageIdMapper implements MessageIdMapper {
+
+    private final MailboxMapper mailboxMapper;
+    private final AttachmentMapper attachmentMapper;
+    private final CassandraMessageIdToImapUidDAO imapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+    private final CassandraMessageDAO messageDAO;
+
+    public CassandraMessageIdMapper(MailboxMapper mailboxMapper, AttachmentMapper attachmentMapper,
+            CassandraMessageIdToImapUidDAO imapUidDAO, CassandraMessageIdDAO messageIdDAO, CassandraMessageDAO messageDAO) {
+        this.mailboxMapper = mailboxMapper;
+        this.attachmentMapper = attachmentMapper;
+        this.imapUidDAO = imapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+        this.messageDAO = messageDAO;
+    }
+
+    @Override
+    public List<Message> find(List<MessageId> messageIds, FetchType fetchType) {
+        List<CassandraMessageId> cassandraMessageIds = messageIds.stream()
+            .map(id -> (CassandraMessageId) id)
+            .collect(Collectors.toList());
+        return messageDAO.retrieveMessages(cassandraMessageIds, fetchType, Optional.empty()).join()
+                .map(pair -> Pair.of(pair.getLeft(), new AttachmentLoader(attachmentMapper).getAttachments(pair.getRight().collect(Guavate.toImmutableList()))))
+                .map(Throwing.function(pair -> {
+                    return SimpleMailboxMessage.cloneWithAttachments(pair.getLeft(), 
+                            pair.getRight().collect(Guavate.toImmutableList()));
+                }))
+                .sorted(Comparator.comparing(MailboxMessage::getUid))
+                .collect(Guavate.toImmutableList());
+    }
+
+    @Override
+    public List<MailboxId> findMailboxes(MessageId messageId) {
+        return imapUidDAO.retrieve((CassandraMessageId) messageId, Optional.empty()).join()
+            .map(ComposedMessageId::getMailboxId)
+            .collect(Guavate.toImmutableList());
+    }
+
+    @Override
+    public void save(MailboxMessage mailboxMessage) throws MailboxNotFoundException, MailboxException {
+        CassandraId mailboxId = (CassandraId) mailboxMessage.getMailboxId();
+        messageDAO.save(mailboxMapper.findMailboxById(mailboxId), mailboxMessage).join();
+        CassandraMessageId messageId = (CassandraMessageId) mailboxMessage.getMessageId();
+        CompletableFuture.allOf(imapUidDAO.insert(messageId, mailboxId, mailboxMessage.getUid()),
+                messageIdDAO.insert(mailboxId, mailboxMessage.getUid(), messageId))
+            .join();
+    }
+
+    @Override
+    public void delete(MessageId messageId) {
+        CassandraMessageId cassandraMessageId = (CassandraMessageId) messageId;
+        messageDAO.delete(cassandraMessageId).join();
+        imapUidDAO.retrieve(cassandraMessageId, Optional.empty()).join()
```

Comment:
I can't, because i'm doing some Cassandra request in `deleteIds` and CassandraDriver doesn't support doing a request in a CompletableFuture.